api/graphql/types/recurring_reservation/serializers.py

```diff
@@ -38,52 +59,316 @@ class Meta:
         for weekday in weekdays:
-            if weekday not in WEEKDAYS.VALUES:
-                raise ValidationErrorWithCode(f"Invalid weekday {weekday}.", ValidationErrorCodes.INVALID_WEEKDAY)
-            days.append(str(weekday))
+            if weekday not in range(6):
+                msg = f"Invalid weekday: {weekday}."
+                raise ValidationError(msg, code=error_codes.RESERVATION_SERIES_INVALID_WEEKDAY)
 
-        return ",".join(days)
+        return ",".join(str(day) for day in weekdays)
 
     def validate_recurrence_in_days(self, recurrence_in_days: int) -> int:
         if recurrence_in_days == 0 or recurrence_in_days % 7 != 0:
-            raise ValidationErrorWithCode(
-                "Interval value not allowed, allowed values are 7,14,28 etc. divisible by seven.",
-                ValidationErrorCodes.INVALID_RECURRENCE_IN_DAY,
-            )
+            msg = "Reoccurrence interval must be a multiple of 7 days."
+            raise ValidationError(msg, code=error_codes.RESERVATION_SERIES_INVALID_RECURRENCE_IN_DAYS)
+
         return recurrence_in_days
+
+    @staticmethod
+    def validate_start_interval(reservation_unit: ReservationUnit, begin_time: datetime.time) -> None:
+        interval_minutes = ReservationStartInterval(reservation_unit.reservation_start_interval).as_number
+
+        # Staff reservations ignore start intervals longer than 30 minutes
+        if interval_minutes != 15:
+            interval_minutes = 30
+
+        # For staff reservations, we don't need to care about opening hours,
+        # so we can just check start interval from the beginning of the day.
+        for hour in range(24):
+            for minute in range(0, 60, interval_minutes):
+                start_time = datetime.time(hour=hour, minute=minute, tzinfo=DEFAULT_TIMEZONE)
+                if start_time == begin_time:
+                    return
+
+        msg = f"Reservation start time does not match the allowed interval of {interval_minutes} minutes."
+        raise ValidationError(msg, code=error_codes.RESERVATION_TIME_DOES_NOT_MATCH_ALLOWED_INTERVAL)
+
+
+class RecurringReservationUpdateSerializer(RecurringReservationCreateSerializer):
+    instance: RecurringReservation
+
+    weekdays = serializers.ListField(child=serializers.IntegerField(), allow_empty=False, required=False)
+
+    class Meta(RecurringReservationCreateSerializer.Meta):
+        extra_kwargs = {
+            "begin_date": {"required": False},
+            "end_date": {"required": False},
+            "begin_time": {"required": False},
+            "end_time": {"required": False},
+            "reservation_unit": {"read_only": False},
+            "recurrence_in_days": {"required": False},
+        }
+
+
+class ReservationSeriesSerializer(RecurringReservationCreateSerializer):
+    """Create the recurring reservation with all its reservations."""
+
+    reservation_type = serializers.ChoiceField(
```

Comment:
Also need to be able to supply other reservation details beside their type (optional, only the type is mandatory) -- also includes buffer time values, cf. https://github.com/City-of-Helsinki/tilavarauspalvelu-core/blob/main/api/graphql/types/reservation/serializers/staff_create_serializers.py#L30

Reply:
These fields are not included in the staff admin create, but are in the model: `sku`, `cancel_details`, `handling_details`, `price`, `price_net`,  `non_subsidised_price`, `non_subsidised_price_net`, `tax_percentage_value`, `deny_reason`, and `cancel_reason`. `user` is also not included, it always uses request user, which I don't think is correct.

Reply:
Yes, these are not needed (either possibly set at update of individual reservation later or in the case of the price stuff not relevant for reservations created by stuff, at least for now): sku, cancel_details, handling_details, price, price_net, non_subsidised_price, non_subsidised_price_net, tax_percentage_value, deny_reason, and cancel_reason.
Other than those, `user` can indeed be supplied in the payload.

Reply:
For some reason, `unit_price` is updateable (only price field), is this correct?

Reply:
No, `unit_price ` shouldn't be updateable.

Reply:
Removed this.